File: 백준/Gold/15591. MooTube （Silver）/MooTube （Silver）.py
import sys
input = sys.stdin.readline
from collections import deque

N, Q = map(int, input().split())
G = [[] for _ in range(N+1)]


for _ in range(N-1):
    p, q, r = map(int, input().split())
    G[p].append((q, r))
    G[q].append((p, r))

for _ in range(Q):
    k, v = map(int, input().split())
    cnt = 0
    deq = deque()
    deq.append((v, 10**9))
    visited = [10**9]*(N+1)

    while deq:
        ver, r = deq.popleft()
        for item in G[ver]: # 인접한 점들에 대하여
            nxt, usado = item # 인접한 점, 현재 점(ver)과 인접한 점(nxt)과의 유사도(usado)
            if nxt != v and visited[nxt] == 10**9: # 아직 방문을 안 했다면
                visited[nxt] = min(r, usado)
                deq.append((nxt, visited[nxt]))
                if visited[nxt] >= k:
                    cnt += 1

    print(cnt)

File: 백준/Gold/15591. MooTube （Silver）/MooTube （Silver）.js
const fs = require('fs')
const input = fs.readFileSync(process.platform === 'linux'? '/dev/stdin' : "./15591input.txt").toString().trim().split("\n").map((el)=>el.split(" ").map(Number))

const [N, Q] = input.shift()
const G = Array.from({ length: N+1 }, () => [])

for (let i = 0; i<N-1; i++) {
  const [p, q, r] = input.shift()
  G[p].push([q, r])
  G[q].push([p, r])
}


input.map( el => {
  const [k, v] = el
  const que = [[v, 10**9]]
  const visited = Array.from({length: N+1}, () => 10**9)
  let cnt = 0
  
  while (que.length > 0) {
    const [cur, r] = que.shift()
    G[cur].map(el => {
      const [nxt, usado] = el
      if (nxt !== v && visited[nxt] === 10**9) {
        visited[nxt] = Math.min(r, usado)
        que.push([nxt, visited[nxt]])
        if (visited[nxt] >= k) {
          cnt += 1
        }
      }
    })
  }
  console.log(cnt)
})